### PR TITLE
pairing: Adds support for BLS12381 using CIRCL library

### PR DESCRIPTION
**Pull request by armfazh**
This PR adds the functionality of BLS12381 pairings using the CIRCL implementation as backend.

[CIRCL](https://github.com/cloudflare/circl/) implementation uses formally-verified implementation (using fiat-crypto) for the prime field arithmetic.

cc: @thibmeu @nikkolasg

### CODE REVIEW

**Review by nikkolasg — COMMENTED** [review]
Really cool ! Thanks a lot for making this !!!
@AnomalRoil @CluEleSsUK there is one thing im wondering is if this should live inside kyber or be its separate repo so kyber stays mostly the interface definition ? I'm not sure.

**Comment by thibmeu** [comment]
Great job! I confirm this does not break tests. It does not replace internal use of `github.com/drand/kyber-bls12381` either.

In term of benchmark, I end up with the following on my machine

The current implementation of drand/kyber-bls12381 using `go test -bench=.`
```
goos: linux
goarch: amd64
pkg: github.com/drand/kyber-bls12381
BenchmarkPairingSeparate-20    	    1026	   1096506 ns/op
BenchmarkPairingInv-20         	    1720	    724972 ns/op
```

This commit implementation armfazh/add_cirl_drand using `go test -bench=. ./pairing/bls12381`
```
goos: linux
goarch: amd64
pkg: github.com/drand/kyber/pairing/bls12381
BenchmarkPairingSeparate-20          310           3983837 ns/op
BenchmarkPairingInv-20               442           2599874 ns/op
```

There is about a 4x difference between the two. I'm not sure if there is a safer/more portable implementation which might explain the difference.

Given there is no modification of actual usage (ibe implementation is not modified for instance), the commit is not going to impact downstream dependencies.

File: pairing/bls12381/adapter.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+// SuiteBLS12381 is an adapter that implements the suites.Suite interface so that
+// bls12381 can be used as a common suite to generate key pairs for instance but
+// still preserves the properties of the pairing (e.g. the Pair function).
+//
+// It's important to note that the Point function will generate a point
+// compatible with public keys only (group G2) where the signature must be
+// used as a point from the group G1.
+type SuiteBLS12381 struct {
```
[the remark this answers is not in the document]
Comment (armfazh) [reply pairing/bls12381/adapter.go:14]: it is enough, I was just mimicking what is already done in the other pairing package.

File: pairing/bls12381/g1.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	ss, qq := s.(*Scalar), q.(*G1Elt)
+	p.inner.ScalarMult(&ss.inner, &qq.inner)
+	return p
+}
+
+func (p *G1Elt) IsInCorrectGroup() bool { return p.inner.IsOnG1() }
+
+func (p *G1Elt) Hash(msg []byte) kyber.Point       { p.inner.Hash(msg, nil); return p }
+func (p *G1Elt) Hash2(msg, dst []byte) kyber.Point { p.inner.Hash(msg, dst); return p }
```
[the remark this answers is not in the document]
Comment (armfazh) [reply pairing/bls12381/g1.go:95]: Once #50 get fixed, there will be only one (correct) API for hashes.

**Comment by armfazh** [comment]
> There is about a 4x difference between the two. I'm not sure if there is a safer/more portable implementation which might explain the difference.

Regarding performance, one of the reasons of this difference is that the kilic library uses assembly for the prime field arithmetic, while CIRCL uses formally-verified code.

**Comment by AnomalRoil** [comment]
Given the way Kyber abstract stuff I think it would make sense to have a `kyber-circle` repo just like we have a [kyber-bls12381](https://github.com/drand/kyber-bls12381) repo that does the interface between kyber and the underlying implementation of the curve.

WDYT?

**Comment by AnomalRoil** [comment]
Not entirely. We are going to be working on merging back drand/kyber into dedis/kyber for the v4 release of dedis/kyber.

However, I have started looking into maintaining a fork of kilic/bls12-381 but the more I dig into it the more bugs I find, so yeah, switching to circl_bls12-381 seems like a plausible option. We'll just need to do some more benchmarking and profiling since we'd really want to be able to run a ~30 node network under the 1s period and need to double check how that would behave wrt aggregation times and so.
If you want to help, it'd be great to have aggregation time benchmarks for various size of polynomials 😇

**Comment by armfazh** [comment]
News: BLS signatures are supported in CIRCL (See [bls.go](https://github.com/cloudflare/circl/blob/main/sign/bls/bls.go)). 
Let us know if more functionality is required.

**Comment by AnomalRoil** [comment]
The initial goal of Kyber was to have all operations implemented in pure go, so having these kind of wrappers built-in the library is already a bit much, so I don't think the goal is to move away from the current Kyber implementation of primitives other than the elliptic curves anytime soon

**Pull request merged**